### PR TITLE
EQ: Add to FIR and IIR support for timer based scheduling

### PR DESCRIPTION
**Pull request by singalsu**
This patch updates the component buffers access for FIR and IIR
equalizer components. Timer based scheduling requires true circular
buffers capability. The copy functions are simplified with a new
utility function comp_copy_helper() that is added to component.c.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

### CODE REVIEW

**Comment by singalsu** [comment]
Note: I need to fix separately later the intrinsic optimized version of the code. I published this because it fixes the distorted sound with gcc and the FW still builds OK with xt-xcc. Xt-xcc build version however remains to sound distorted due to not yet treating the circular component buffers correctly.

File: src/audio/component.c [hunk 1]
```diff
@@ -239,3 +239,36 @@ void sys_comp_init(void)
 		((void(*)(void))(*comp_init))();
 	}
 }
+
+int comp_copy_helper(struct comp_dev *dev, struct comp_copy_util *u)
```
Comment (ranj063) [inline src/audio/component.c]: @singalsu this could probably be a separate patch and applied to other comps that do the same thing (volume, selector etc)

File: src/audio/eq_fir.c [hunk 2]
```diff
@@ -692,33 +675,42 @@ static int eq_fir_prepare(struct comp_dev *dev)
...
-	dev->frame_bytes = cd->period_bytes / dev->frames;
+	/* get sink data format */
+	comp_set_period_bytes(sinkb->sink, dev->frames, &cd->sink_format,
+			      &sink_period_bytes);
+
+	/* Rewrite params format for this component to match the host side. */
+	if (dev->params.direction == SOF_IPC_STREAM_PLAYBACK)
+		dev->params.frame_fmt = cd->source_format;
+	else
+		dev->params.frame_fmt = cd->sink_format;
```
Comment (ranj063) [inline src/audio/eq_fir.c]: @singalsu im wondering if this can be rolled into the comp_set_period_bytes() method. We can probably rename it to something generic like comp_set_copy_params() possibly

File: src/audio/eq_iir.c [hunk 3]
```diff
@@ -147,24 +78,23 @@ static void eq_iir_s16_default(struct comp_dev *dev,
...
-		y = snk++;
+		idx = ch;
 		for (i = 0; i < frames; i++) {
+			x = buffer_read_frag_s16(source, idx);
+			y = buffer_write_frag_s16(sink, idx);
 			z = iir_df2t(filter, *x << 16);
 			*y = sat_int16(Q_SHIFT_RND(z, 31, 15));
-			x += nch;
-			y += nch;
+			idx += nch;
```
Comment (ranj063) [inline src/audio/eq_iir.c]: @singalsu is this correct? Should this be idx = frame *nch + ch?

Comment (ranj063) [reply src/audio/eq_iir.c]: sorry this is correct. I missed the first idx = ch;

**Comment by lgirdwood** [comment]
@singalsu btw, I assume the circular buffers also work fine in test bench. If so we need to look at the code generated by xt-xcc, it may be over optimised and we may need to add some hints to help the xt-xcc compiler.

**Comment by singalsu** [comment]
@lgirdwood The xt-xcc version issue is due to known missing code rework that I need to do next. I didn't have enough time before my vacation to include it too into this PR.

I'll address the comments here with a new PR version soon.

**Comment by lgirdwood** [comment]
@singalsu ok great, please also make sure FIR/IIR will still work well when scheduled via DMA IRQ.

File: src/audio/component.c (hunk 1 above)
Comment (singalsu) [reply src/audio/component.c]: ok!

File: src/audio/eq_fir.c (hunk 2 above)
Comment (singalsu) [reply src/audio/eq_fir.c]: That's not straightforward. The cd->source_format and sink_format are defined in the component locally and there's no common SOF  header for them.  If I split the format discovery from comp_set_period_bytes() into a separate function the function comp_set_copy_params() could use it to get the source/sink format information. I don't know if that would be nicer than this.

Comment (singalsu) [reply src/audio/eq_fir.c]: @ranj063 I'll leave this undone for next PR but if you like the idea I will try to do for next proposal.

**Comment by singalsu** [comment]
I just pushed a new version. There are no code updates except that it's rebased and split to five commits for easier review. Functional EQs require also PR #1191 that I provided as separate PR (not related bug).

File: src/audio/component.c (hunk 1 above)
Comment (lgirdwood) [reply src/audio/component.c]: naming here could be improved tp be more meaningful. comp_get_copy_limits()

File: src/include/sof/audio/component.h [hunk 4]
```diff
@@ -272,6 +272,15 @@ struct comp_dev {
  *  @{
  */
 
+/** \brief Struct for use with comp_copy_helper() function. */
+struct comp_copy_util {
```
Comment (lgirdwood) [inline src/include/sof/audio/component.h]: struct comp_copy_limits

File: src/audio/eq_iir.c [hunk 5]
```diff
@@ -797,33 +793,33 @@ static int eq_iir_prepare(struct comp_dev *dev)
 	if (ret == COMP_STATE_ALREADY_SET)
 		return PPL_PATH_STOP;
 
-	/* EQ components will only ever have 1 source and 1 sink buffer */
+	/* Get EQ source and buffers. */
```
Comment (lgirdwood) [inline src/audio/eq_iir.c]: This will still be true since the SRC is not a mixer or mux

File: src/audio/eq_fir.c [hunk 6]
```diff
@@ -734,33 +717,42 @@ static int eq_fir_prepare(struct comp_dev *dev)
 	if (ret == COMP_STATE_ALREADY_SET)
 		return PPL_PATH_STOP;
 
-	/* EQ components will only ever have 1 source and 1 sink buffer */
+	/* Get EQ source and buffers. */
```
Comment (lgirdwood) [inline src/audio/eq_fir.c]: Ditto for number of sources and sinks.

**Comment by lgirdwood** [comment]
@singalsu CI build/UT also failing

File: src/audio/component.c (hunk 1 above)
Comment (singalsu) [reply src/audio/component.c]: ok!

File: src/audio/eq_iir.c (hunk 5 above)
Comment (singalsu) [reply src/audio/eq_iir.c]: OK, seems like I messed up the updated comment text ("sink" missing) but I'll revert to old since you prefer it.

File: src/include/sof/audio/component.h (hunk 4 above)
Comment (singalsu) [reply src/include/sof/audio/component.h]: Done!

**Comment by singalsu** [comment]
I've just pushed new version of this.

**Comment by singalsu** [comment]
@zrombel Can you help with the CI error. Is it real issue or something in the test system?

**Comment by singalsu** [comment]
The CI issue was for cmocka test for mixer component. I just force pushed the same PR to see if it happens again.

**Comment by singalsu** [comment]
I found something from CI log to check if it has relation to PR.

```
make -j8 VERBOSE=1 && ctest -j16 --verbose
Command return code: 2
Command error output: /quickbuild/workspace/24733/SOF_FW/test/cmocka/src/audio/mixer/comp_mock.c:83: warning: initialization from incompatible pointer type
CMakeFiles/comp_set_state.dir/__/__/__/__/__/src/audio/component.c.o: In function `comp_underrun':
component.c:(.text+0x225): undefined reference to `pipeline_xrun'
CMakeFiles/comp_set_state.dir/__/__/__/__/__/src/audio/component.c.o: In function `comp_overrun':
component.c:(.text+0x269): undefined reference to `pipeline_xrun'
make[2]: *** [test/cmocka/src/audio/component/comp_set_state] Error 2
make[1]: *** [test/cmocka/src/audio/component/CMakeFiles/comp_set_state.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
```

**Comment by singalsu** [comment]
As suggested by @tlauda I modified the PR to include for cmocka an empty function pipeline_xrun() into cmocka/src/audio/component/mock.c. I hope it helps to pass the CI.

**Comment by lgirdwood** [comment]
@singalsu still have a CI issue with code check.

**Comment by singalsu** [comment]
@lgirdwood I see "7 successful checks" above. Do you remember what the issue you saw was? The logs are not easy to text search for e.g. string "error".

**Comment by lgirdwood** [comment]
@singalsu all CI good now thanks !
